compact_str/src/lib.rs

```diff
@@ -667,6 +667,72 @@ impl CompactString {
         );
         unsafe { self.set_len(new_len) };
     }
+
+    /// Converts a [`CompactString`] to a raw pointer.
+    #[inline]
+    pub fn as_ptr(&mut self) -> *const u8 {
+        self.repr.as_slice().as_ptr()
+    }
+
+    /// Converts a mutable [`CompactString`] to a raw pointer.
+    #[inline]
+    pub fn as_mut_ptr(&mut self) -> *mut u8 {
+        unsafe { self.repr.as_mut_slice().as_mut_ptr() }
+    }
```

Comment:
Yes, but going the deref route is more costly, unless the compiler can tell that the temporary &str creation including a checked indexing can be omitted. I haven't looked at the assembly code to see how smart llvm is. :)

Reply:
Great point! I wasn't actually sure if rustc or LLVM were smart enough here, so I looked into it!

First I checked using [godbolt](https://godbolt.org/z/WfaKj1xej) and it looks like they should be the same. I also made a small binary:

```
use compact_str::CompactString;

fn main() {
    let mut compact = CompactString::new("hello world");
    compact.insert_str(4, "blah");
    println!("{}", compact);
}
``` 

And compiled it twice emitting asm, once with our own `fn as_mut_ptr(...)`, and once without. Specifically I ran `cargo rustc --bins --release -- --emit asm` and then ran `cargo clean` between each run.

I was quite surprised to find that the emitted assembly was exactly the same! Pretty neat to see rustc or LLVM can optimize this much :)

Reply:
I think you would have to use [std::hint::black_box(compact)](https://doc.rust-lang.org/std/hint/fn.black_box.html) to get an accurate result. Otherwise the compiler can tell than `compact` is inline, and won't generate code for the other case. I tested it with

```rust
pub fn as_mut_ptr1(&mut self) -> *mut u8 {
    unsafe { self.repr.as_mut_slice().as_mut_ptr() }
}

pub fn as_mut_ptr2(&mut self) -> *mut u8 {
    self.deref_mut().as_mut_ptr()
}
```

```sh
objdump -dC ./target/release/libcompact_str.rlib > generated.S
```
```S
Disassembly of section .text._RNvMCs3n1ol3MNYow_11compact_strNtB2_13CompactString11as_mut_ptr1:

0000000000000000 <<compact_str::CompactString>::as_mut_ptr1>:
   0:	48 83 ec 38          	sub    $0x38,%rsp
   4:	8a 47 17             	mov    0x17(%rdi),%al
   7:	3c fe                	cmp    $0xfe,%al
   9:	74 0c                	je     17 <<compact_str::CompactString>::as_mut_ptr1+0x17>
   b:	3c ff                	cmp    $0xff,%al
   d:	74 13                	je     22 <<compact_str::CompactString>::as_mut_ptr1+0x22>
   f:	48 89 f8             	mov    %rdi,%rax
  12:	48 83 c4 38          	add    $0x38,%rsp
  16:	c3                   	ret    
  17:	48 8b 3f             	mov    (%rdi),%rdi
  1a:	48 89 f8             	mov    %rdi,%rax
  1d:	48 83 c4 38          	add    $0x38,%rsp
  21:	c3                   	ret    
  22:	48 8d 7c 24 08       	lea    0x8(%rsp),%rdi
  27:	e8 00 00 00 00       	call   2c <<compact_str::CompactString>::as_mut_ptr1+0x2c>
  2c:	31 c0                	xor    %eax,%eax
  2e:	48 83 c4 38          	add    $0x38,%rsp
  32:	c3                   	ret    

Disassembly of section .text._RNvMCs3n1ol3MNYow_11compact_strNtB2_13CompactString11as_mut_ptr2:

0000000000000000 <<compact_str::CompactString>::as_mut_ptr2>:
   0:	48 83 ec 38          	sub    $0x38,%rsp
   4:	8a 47 17             	mov    0x17(%rdi),%al
   7:	3c fe                	cmp    $0xfe,%al
   9:	74 0c                	je     17 <<compact_str::CompactString>::as_mut_ptr2+0x17>
   b:	3c ff                	cmp    $0xff,%al
   d:	74 33                	je     42 <<compact_str::CompactString>::as_mut_ptr2+0x42>
   f:	48 89 f8             	mov    %rdi,%rax
  12:	48 83 c4 38          	add    $0x38,%rsp
  16:	c3                   	ret    
  17:	48 8b 47 08          	mov    0x8(%rdi),%rax
  1b:	0f b7 4f 14          	movzwl 0x14(%rdi),%ecx
  1f:	0f b6 57 16          	movzbl 0x16(%rdi),%edx
  23:	c1 e2 10             	shl    $0x10,%edx
  26:	09 ca                	or     %ecx,%edx
  28:	48 c1 e2 20          	shl    $0x20,%rdx
  2c:	8b 77 10             	mov    0x10(%rdi),%esi
  2f:	48 09 d6             	or     %rdx,%rsi
  32:	48 39 f0             	cmp    %rsi,%rax
  35:	77 1c                	ja     53 <<compact_str::CompactString>::as_mut_ptr2+0x53>
  37:	48 8b 3f             	mov    (%rdi),%rdi
  3a:	48 89 f8             	mov    %rdi,%rax
  3d:	48 83 c4 38          	add    $0x38,%rsp
  41:	c3                   	ret    
  42:	48 8d 7c 24 08       	lea    0x8(%rsp),%rdi
  47:	e8 00 00 00 00       	call   4c <<compact_str::CompactString>::as_mut_ptr2+0x4c>
  4c:	31 c0                	xor    %eax,%eax
  4e:	48 83 c4 38          	add    $0x38,%rsp
  52:	c3                   	ret    
  53:	48 8d 15 00 00 00 00 	lea    0x0(%rip),%rdx        # 5a <<compact_str::CompactString>::as_mut_ptr2+0x5a>
  5a:	48 89 c7             	mov    %rax,%rdi
  5d:	ff 15 00 00 00 00    	call   *0x0(%rip)        # 63 <<compact_str::CompactString>::as_mut_ptr2+0x63>
  63:	0f 0b                	ud2    
```

The `self.as_mut_str().as_mut_ptr()` variant generates more code, and also the checked call was not elided. `ud2` causes a SIGILL, and is used to check that `Fn() -> !` actually didn't return. So the compiler does not seem be as smart as one could have hoped. :D

Reply:
Oh oh! Great idea to use `std::hint::black_box(...)` and great analysis! We'll stick with how it is :)